### ISSUE TITLE
CustomHosting in V4 behaves differently to NServiceBus.Host v4 when CustomCheck creation fails

### ISSUE THREAD

**Issue by danielmarbach** [Particular/NServiceBus#2654]
CustomHosting in V4 behaves differently than NServiceBus.Host v4 and CustomHosting in V5 and NServiceBus.Host V5

Here the reproduction steps:

* Use the custom check from [this issue](https://github.com/Particular/ServiceControl.Plugins.v5/issues/8) make the monitor base class non-abstract
* Create console application which host NServiceBus 4.7.5

        Configure.Serialization.Json();
        var configure = Configure.With();
        configure.DefineEndpointName("Samples.CustomChecks.Monitor3rdParty");
        configure.DefaultBuilder();
        configure.InMemorySagaPersister();
        configure.UseInMemoryTimeoutPersister();
        configure.InMemorySubscriptionStorage();
        configure.UseTransport<Msmq>();
        var bus = configure.UnicastBus()
            .CreateBus()
            .Start(() => Configure.Instance.ForInstallationOn<Windows>().Install());

        Console.WriteLine("\r\nPress any key to stop program\r\n");
        Console.ReadKey();

* Host the custom check in that endpoint
* Start the endpoint
* See `Press any key to stop program`

Expected behavior
* The endpoint should not start and throw an activation exception
* The bus instance should be shutdown

### Background information

Custom Host V5, NServiceBus.Host V5 and NServiceBus.Host V4 behave correctly by shutting down the bus and throwing an exception.

// cc @SeanFeldman

**Comment by danielmarbach** [comment]
The issue is that the PeriodicCheckMonitorTask remains in Running state forever. Therefore blocking the shutdown of the bus inside Shutdown where we wait for `thingsToRunAtStartupTask` to complete.

**Comment by danielmarbach** [comment]
I think I got it. It seems that the commit a9c57eb6c0d64bd52aab539776fccba0da33ac34 hasn't been included in 4.7.5. See

![deadlockcriticalaction](https://cloud.githubusercontent.com/assets/174258/6169494/1a2e6716-b2ce-11e4-9ee6-9dfd407ebda4.jpg)

So in 4.7.5 we are essentially dead locking during shutdown.

**Comment by danielmarbach** [comment]
I cherry picked that commit to the hotfix-4.7.6 branch see 8aa5cd48fd8a459b03372b7a734426055e258dfb

**Comment by danielmarbach** [comment]
So this fixed the dead lock issue. But it doesn't fix the issue of the activation exception being swallowed

**Comment by johnsimons** [comment]
> But it doesn't fix the issue of the activation exception being swallowed

@danielmarbach what version is this exception being swallowed ?

**Comment by johnsimons** [comment]
ping @danielmarbach

**Comment by danielmarbach** [comment]
@johnsimons pong ;)

Here the repro. Install 4.7.5 and write the following program:

	class Program
    {
        static void Main(string[] args)
        {
            Configure.Serialization.Json();
            var configure = Configure.With();
            configure.DefineEndpointName("Samples.CustomChecks.Monitor3rdParty");
            configure.DefaultBuilder();
            configure.InMemorySagaPersister();
            configure.UseInMemoryTimeoutPersister();
            configure.InMemorySubscriptionStorage();
            configure.UseTransport<Msmq>();
            var bus = configure.UnicastBus()
                .CreateBus()
                .Start(() => Configure.Instance.ForInstallationOn<Windows>().Install());

            bus.SendLocal(new Message());

            Console.WriteLine("\r\nPress any key to stop program\r\n");

            Console.ReadKey();
        }

        public class Runner : IWantToRunWhenBusStartsAndStops
        {
            public void Start()
            {
                throw new InvalidOperationException();
            }

            public void Stop()
            {
            }
        }
    }

    internal class Message : ICommand
    {
    }

**Comment by danielmarbach** [comment]
The exceptions are also swallowed in 4.7.4. and 4.7.3

**Comment by danielmarbach** [comment]
The issue is the following:

In contrast to `master` when a `IWantToRunWhenTheBusStartsAndStops` throws an exception the exception is catched inside the tasks and swallowed (from a task perspective). This means the task never goes to the faulted state. Therefore when we wait on those tasks the exception is never thrown to the caller. Is this expected? In V5 it is different. The exception is propagated to the user.

**Comment by johnsimons** [comment]
@danielmarbach I'm on the fence to fix this, what does the rest of @Particular/nservicebus  think ?

**Comment by danielmarbach** [comment]
@johnsimons I understand that. But I think it is misleading that when an important part of the infrastructure throws the custom hosting remains running.

**Comment by johnsimons** [comment]
@danielmarbach it is my understanding that we are fixing this issue as part of v6?
Are you proposing we backport a fix?

**Comment by danielmarbach** [comment]
Not sure if it is worth the effort of backporting. We are not losing any messages so I'd say we leave

**Comment by johnsimons** [comment]
> We are not losing any messages so I'd say we leave

So are you happy with closing this as won't fix ?
